ReactiveCocoa/Swift/Signal.swift

```diff
@@ -299,24 +299,105 @@ extension SignalType {
 	}
 }
 
-/// A reference type which wraps an array to avoid copying it for performance and
-/// memory usage optimization.
-private final class CollectState<Value> {
-	var values: [Value] = []
-
-	func append(value: Value) -> Self {
-		values.append(value)
-		return self
-	}
+public enum CollectStrategy {
+	case Inclusive
+	case Exclusive
 }
 
 extension SignalType {
+
 	/// Returns a signal that will yield an array of values when `self` completes.
 	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
 	public func collect() -> Signal<[Value], Error> {
-		return self
-			.reduce(CollectState()) { $0.append($1) }
-			.map { $0.values }
+		return collect(.Inclusive) { _ in false }
+	}
+
+	/// Returns a signal that will yield an array of values until it reaches a certain
+	/// count.
+	///
+	/// When the count is reached the array is sent and the signal starts over 
+	/// yielding a new array of values.
+	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
+	public func collect(count: Int) -> Signal<[Value], Error> {
+		precondition(count > 0)
+		return collect(.Inclusive) { values, _ in values.count == count }
+	}
+
+	/// Returns a signal that will yield an array of values based on a predicate.
+	///
+	/// The predicate should return `true` when the values should be sent and `false`
+	/// when the values should be collected. The predicate receives the `values`
+	/// already collected and the next value that should or should not be
+	/// collected.
+	///
+	/// #### Example
+	///
+	///     let (signal, observer) = Signal<Int, NoError>.pipe()
+	///
+	///     signal
+	///         .collect { values, next in next == 7 }
+	///         .observeNext { print($0) }
+	///
+	///     observer.sendNext(1)
+	///     observer.sendNext(1)
+	///     observer.sendNext(7)
+	///     observer.sendNext(7)
+	///     observer.sendNext(5)
+	///     observer.sendNext(6)
+	///     observer.sendCompleted()
+	///
+	///     // Output:
+	///     // [1, 1]
+	///     // [7]
+	///     // [7, 5, 6]
+	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
+	public func collect(strategy: CollectStrategy, predicate: (values: [Value], next: Value) -> Bool) -> Signal<[Value], Error> {
```

Comment:
Okay, I think we should split this into two methods:

```swift
// inclusive version
public func collect(predicate: (values: [Value]) -> Bool) -> Signal<[Value], Error>

// exclusive version
public func collect(predicate: (values: [Value], next: Value) -> Bool) -> Signal<[Value], Error> 
```

The existing implementation here treats `values` different depending on whether it's inclusive or exclusive. Iff it's inclusive `value.contains(next)`. I think that sort of subtlety is a mistake, but I understand why you did it. If it's inclusive, you want to that to be true.

So I think we should split it into two different APIs. That would address all my reservations about this.